[scroll-animations] Implement parsing support for animation-range and its longhand properties

nmoucht opened the pull request:
#### 0b69c4397a4c4be00d3ede312efc7a0a9ab3aeba
<pre>
[scroll-animations] Implement parsing support for animation-range and its longhand properties
<a href="https://bugs.webkit.org/show_bug.cgi?id=280862">https://bugs.webkit.org/show_bug.cgi?id=280862</a>
<a href="https://rdar.apple.com/137717240">rdar://137717240</a>

Reviewed by Tim Nguyen and Darin Adler.

Implement parsing support for animation-range and its longhand properties. This needs a follow up
for sequences of animation ranges. There is also a failing subtest about determining if a calc
expression is a default offset value that needs some thought.

* Source/WebCore/Headers.cmake:
* Source/WebCore/animation/AnimationTimeline.h:
(WebCore::SingleTimelineRange::operator== const):
* Source/WebCore/animation/TimelineRangeOffset.h:
(WebCore::TimelineRangeOffset::operator== const):
* Source/WebCore/animation/WebAnimationTypes.h:
* Source/WebCore/css/CSSProperties.json:
* Source/WebCore/css/CSSValueKeywords.in:
* Source/WebCore/css/ComputedStyleExtractor.cpp:
(WebCore::valueForAnimationRange):
(WebCore::ComputedStyleExtractor::valueForPropertyInStyle const):
* Source/WebCore/css/parser/CSSPropertyParserConsumer+Timeline.cpp:
(WebCore::CSSPropertyParserHelpers::consumeAnimationRange):
* Source/WebCore/css/parser/CSSPropertyParserConsumer+Timeline.h:
* Source/WebCore/rendering/style/RenderStyle.h:
* Source/WebCore/rendering/style/RenderStyleInlines.h:
(WebCore::RenderStyle::initialAnimationRangeStart):
(WebCore::RenderStyle::initialAnimationRangeEnd):
(WebCore::RenderStyle::animationRangeStart const):
(WebCore::RenderStyle::animationRangeEnd const):
* Source/WebCore/rendering/style/RenderStyleSetters.h:
(WebCore::RenderStyle::setAnimationRangeStart):
(WebCore::RenderStyle::setAnimationRangeEnd):
* Source/WebCore/rendering/style/StyleRareNonInheritedData.cpp:
(WebCore::StyleRareNonInheritedData::StyleRareNonInheritedData):
(WebCore::StyleRareNonInheritedData::operator== const):
* Source/WebCore/rendering/style/StyleRareNonInheritedData.h:
* Source/WebCore/style/StyleBuilderConverter.h:
(WebCore::Style::BuilderConverter::convertAnimationRangeStart):
(WebCore::Style::BuilderConverter::convertAnimationRangeEnd):
(WebCore::Style::BuilderConverter::convertAnimationRange):

Canonical link: <a href="https://commits.webkit.org/285135@main">https://commits.webkit.org/285135@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/86778a6172eb09b40ab140018d080f86a6b8b5b2

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/71679 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/51092 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/24453 "Built successfully") | [  ~~🛠 wpe~~](https://ews-build.webkit.org/#/builders/5/builds/75793 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/22884 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/73794 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/58893 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/22704 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/5/builds/75793 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/60/builds/15096 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/74745 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/47/builds/46357 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/61748 "Passed tests") | [  ~~🧪 api-wpe~~](https://ews-build.webkit.org/#/builders/5/builds/75793 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/43018 "Passed tests") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/64/builds/19215 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/21225 "Built successfully") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/64923 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/19578 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/77513 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/87/builds/15913 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/62/builds/18761 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/64324 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/86/builds/15957 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/61781 "Passed tests") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/21/builds/64325 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/15834 "Built successfully and passed tests") | [  ~~🧪 vision-wk2~~](https://ews-build.webkit.org/#/builders/88/builds/12481 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-intel-wk2~~](https://ews-build.webkit.org/#/builders/119/builds/6120 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/46892 "Built successfully") | [  ~~🛠 mac-safer-cpp~~](https://ews-build.webkit.org/#/builders/120/builds/1671 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/47963 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/49247 "Built successfully") | | | 
| | [  ~~🛠 watch-sim~~](https://ews-build.webkit.org/#/builders/45/builds/47705 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
<!--EWS-Status-Bubble-End-->

smfr commented on Source/WebCore/css/parser/CSSPropertyParserConsumer+Timeline.cpp:
```diff
@@ -150,5 +152,41 @@ RefPtr<CSSValue> consumeViewTimelineInset(CSSParserTokenRange& range, const CSSP
+    auto id = range.peek().id();
+
+    if (id == CSSValueNormal)
+        return consumeIdent(range);
```
You may need to check that you're at the end of the range after consuming the ident, to ensure that something like "normal foo" or "normal 10px" is a parsing error. The tests should have some invalid value tests, which you should add to if they don't include all the invalid configurations.

nt1m replied:
This isn't really needed, this is called from `parseValueStart` or `parseSingleValue` which have their own `range.atEnd()` checks.

I do agree some tests for invalid values would be nice though.

nmoucht replied:
Right its not necessary to add a check. Did add some extra invalid tests though.

smfr commented on Source/WebCore/animation/TimelineRange.h:
```diff
@@ -29,9 +29,22 @@
 
 namespace WebCore {
 
+enum class TimelineRangeStartOrEnd { Start, End };
```
`: bool`

nmoucht replied:
Fixed.

smfr commented on Source/WebCore/animation/TimelineRangeOffset.h:
```diff
@@ -33,6 +33,11 @@ namespace WebCore {
     String rangeName;
     RefPtr<CSSNumericValue> offset;
+
+    bool operator==(const TimelineRangeOffset& rangeOffset) const
```
Maybe just ` = default` ?

nmoucht replied:
Fixed.

nt1m commented on Source/WebCore/css/CSSProperties.json:
```diff
@@ -1557,6 +1557,45 @@
+                "converter": "AnimationRangeStart",
+                "parser-function": "consumeAnimationRangeStart",
+                "parser-grammar-unused": "[ normal | <length-percentage> | [ <custom-ident> || <length-percentage>?] ]#",
+                "parser-grammar-unused-reason": "Needs support for function notation.",
```
This comment is probably wrong.

Maybe `"parser-grammar-unused-reason": "Needs support for '||' groups."`

nmoucht replied:
Fixed.

nt1m commented on Source/WebCore/css/CSSProperties.json:
```diff
@@ -1557,6 +1557,45 @@
+                "converter": "AnimationRangeEnd",
+                "parser-function": "consumeAnimationRangeEnd",
+                "parser-grammar-unused": "[ normal | <length-percentage> | [ <custom-ident> || <length-percentage>? ] ]#",
+                "parser-grammar-unused-reason": "Needs support for function notation.",
```
ditto

nmoucht replied:
Fixed.

nt1m commented on Source/WebCore/style/StyleBuilderConverter.h:
```diff
@@ -2187,6 +2191,29 @@ inline TimelineScope BuilderConverter::convertTimelineScope(BuilderState&, const
+inline SingleTimelineRange BuilderConverter::convertAnimationRange(BuilderState& state, const CSSValue& value, TimelineRangeStartOrEnd startOrEnd)
+{
+    if (auto* primitiveValue = dynamicDowncast<CSSPrimitiveValue>(value)) {
+        if (primitiveValue->valueID() == CSSValueInvalid)
```
Maybe check for normal here, and handle that first?

Then assert `primitiveValue->isLength() || primitiveValue->isPercentage()` for the other case

nmoucht replied:
Not sure what you mean by check for normal. I did add these asserts though.

nt1m commented on Source/WebCore/style/StyleBuilderConverter.h:
```diff
@@ -2187,6 +2191,29 @@ inline TimelineScope BuilderConverter::convertTimelineScope(BuilderState&, const
+inline SingleTimelineRange BuilderConverter::convertAnimationRange(BuilderState& state, const CSSValue& value, TimelineRangeStartOrEnd startOrEnd)
+{
+    if (auto* primitiveValue = dynamicDowncast<CSSPrimitiveValue>(value)) {
+        if (primitiveValue->valueID() == CSSValueInvalid)
```
Also, in front of every case, can you add a comment showing the syntax it corresponds to?

```
// <length-percentage>

// normal

// <timeline-range-name> <length-percentage>?
```

nmoucht replied:
Fixed.

nt1m commented on Source/WebCore/css/ShorthandSerializer.cpp:
```diff
@@ -1315,6 +1318,51 @@ String ShorthandSerializer::serializeTextWrap() const
+    if (rangeStart.valueID() == CSSValueNormal && rangeEnd.valueID() == CSSValueNormal)
+        return nameLiteral(CSSValueNormal);
+
+    if (auto* valuePair = dynamicDowncast<CSSValuePair>(rangeStart)) {
```
nit: I usually just name those variables `pair`

nmoucht replied:
Fixed.

nt1m commented on Source/WebCore/css/ShorthandSerializer.cpp:
```diff
@@ -1315,6 +1318,51 @@ String ShorthandSerializer::serializeTextWrap() const
+            rangeStartValue = nameLiteral(valuePair->first().valueID());
+        else
+            rangeStartValue = serializeLonghandValue(0);
+    } else if (auto* valuePrimitive = dynamicDowncast<CSSPrimitiveValue>(rangeStart)) {
```
and this `primitiveValue`

nmoucht replied:
Fixed.

nt1m commented on Source/WebCore/css/ShorthandSerializer.cpp:
```diff
@@ -1315,6 +1318,51 @@ String ShorthandSerializer::serializeTextWrap() const
+            rangeStartValue = nameLiteral(rangeStart.valueID());
+    }
+
+    if (auto* valuePair = dynamicDowncast<CSSValuePair>(rangeEnd)) {
```
ditto

nmoucht replied:
Fixed.

nt1m commented on Source/WebCore/css/ShorthandSerializer.cpp:
```diff
@@ -1315,6 +1318,51 @@ String ShorthandSerializer::serializeTextWrap() const
+            rangeEndValue = nameLiteral(valuePair->first().valueID());
+        else
+            rangeEndValue = serializeLonghandValue(1);
+    } else if (auto* valuePrimitive = dynamicDowncast<CSSPrimitiveValue>(rangeEnd)) {
```
ditto

nmoucht replied:
Fixed.

nt1m commented on Source/WebCore/css/parser/CSSPropertyParser.cpp:
```diff
@@ -2771,6 +2771,41 @@ bool CSSPropertyParser::consumeWhiteSpaceShorthand(bool important)
     return true;
 }
 
+static RefPtr<CSSValue> rangeEndValueForCSSValueID(CSSValueID valueID)
```
I think this method could be more clear (and potentially just a lambda), `rangeEndValueForStartValue` could be a good name.

If there's a way to write this more similarly to the spec:
> If [<'animation-range-end'>](https://drafts.csswg.org/scroll-animations-1/#propdef-animation-range-end) is omitted and [<'animation-range-start'>](https://drafts.csswg.org/scroll-animations-1/#propdef-animation-range-start) includes a [<timeline-range-name>](https://drafts.csswg.org/scroll-animations-1/#typedef-timeline-range-name) component, then animation-range-end is set to that same <timeline-range-name> and 100%. Otherwise, any omitted [longhand](https://drafts.csswg.org/css-cascade-5/#longhand) is set to its [initial value](https://drafts.csswg.org/css-cascade-5/#initial-value).

it would be good

nmoucht replied:
Fixed.

nt1m commented on Source/WebCore/css/ComputedStyleExtractor.cpp:
```diff
@@ -3207,6 +3209,36 @@ static Ref<CSSValue> valueForTimelineScopeNames(const Vector<AtomString>& names)
+    auto rangeStart = style.animationRangeStart();
+    auto rangeEnd = style.animationRangeEnd();
+
+    RefPtr listValueStart = dynamicDowncast<CSSValueList>(valueForSingleAnimationRange(style, rangeStart, TimelineRangeStartOrEnd::Start));
```
startValue

nmoucht replied:
Fixed.

nt1m commented on Source/WebCore/css/ComputedStyleExtractor.cpp:
```diff
@@ -3207,6 +3209,36 @@ static Ref<CSSValue> valueForTimelineScopeNames(const Vector<AtomString>& names)
+    if (listValueStart && listValueStart->length())
+        list.append(*listValueStart);
+
+    RefPtr listValueEnd = dynamicDowncast<CSSValueList>(valueForSingleAnimationRange(style, style.animationRangeEnd(), TimelineRangeStartOrEnd::End));
```
endValue

nmoucht replied:
Fixed.

darinadler commented on Source/WebCore/animation/TimelineRange.cpp:
```diff
@@ -0,0 +1,113 @@
+        return CSSValueEntryCrossing;
+    case SingleTimelineRange::Name::ExitCrossing:
+        return CSSValueExitCrossing;
+    default:
```
It's better to omit `default` in a `switch` like this one so we get a warning if we omit a `case`.

nmoucht replied:
Fixed.

darinadler commented on Source/WebCore/animation/TimelineRange.h:
```diff
@@ -29,9 +29,25 @@
+enum class TimelineRangeStartOrEnd : bool { Start, End };
+
 struct SingleTimelineRange {
+
```
Omit this blank line.

nmoucht replied:
Fixed.

darinadler commented on Source/WebCore/animation/TimelineRange.h:
```diff
@@ -29,9 +29,25 @@
+
 struct SingleTimelineRange {
+
+    bool operator==(const SingleTimelineRange& other) const
```
Compiler will generate this for us if we use `= default`.
```
bool operator==(const SingleTimelineRange&) = default;
```
And this pull request uses that elsewhere. Unclear why not using it here.

nmoucht replied:
Fixed.

darinadler commented on Source/WebCore/animation/TimelineRange.h:
```diff
@@ -29,9 +29,25 @@
+
+    static bool isOffsetValue(const CSSPrimitiveValue&);
+
+    static Name cssPropertyIdToNamedTimelineRange(CSSValueID);
```
Since these are member functions we can give them much shorter names, like just `name`. Argument types are part of a function signature, although return types are not.

nmoucht replied:
Fixed.

darinadler commented on Source/WebCore/animation/TimelineRange.h:
```diff
@@ -29,9 +29,25 @@
+    static bool isOffsetValue(const CSSPrimitiveValue&);
+
+    static Name cssPropertyIdToNamedTimelineRange(CSSValueID);
+    static CSSValueID namedTimelineRangeToCSSPropertyId(Name);
```
Should capitalize `ID`. A value ID and a property ID are two different things, so this should not call it a property ID. Since these are member functions we can give them much shorter names, like just `valueID`.

nmoucht replied:
Fixed.

darinadler commented on Source/WebCore/animation/TimelineRange.h:
```diff
@@ -29,9 +29,25 @@
+
     enum class Name { Normal, Omitted, Cover, Contain, Entry, Exit, EntryCrossing, ExitCrossing };
 
+    static bool isTimelineRangeDefaultValue(const Length&, TimelineRangeStartOrEnd);
```
No need to include the type name. Argument types are part of the function signature. Just name this `isDefault`.

nmoucht replied:
Fixed.

darinadler commented on Source/WebCore/animation/TimelineRange.h:
```diff
@@ -29,9 +29,25 @@
     enum class Name { Normal, Omitted, Cover, Contain, Entry, Exit, EntryCrossing, ExitCrossing };
 
+    static bool isTimelineRangeDefaultValue(const Length&, TimelineRangeStartOrEnd);
+    static bool isTimelineRangeDefaultValue(const CSSPrimitiveValue&, TimelineRangeStartOrEnd);
```
No need to include the type name. Argument types are part of the function signature. Just name this `isDefault`.

nmoucht replied:
Fixed.

darinadler commented on Source/WebCore/animation/TimelineRange.h:
```diff
@@ -29,9 +29,25 @@
 
+enum class TimelineRangeStartOrEnd : bool { Start, End };
+
 struct SingleTimelineRange {
```
We normally put data members first in a `struct` and should do that here.

nmoucht replied:
Fixed.

darinadler commented on Source/WebCore/animation/WebAnimationTypes.h:
```diff
@@ -29,6 +29,7 @@
 #include "CSSPropertyNames.h"
 #include "CSSValue.h"
 #include "EventTarget.h"
+#include "TimelineRangeOffset.h"
```
It's best to minimize the amount that headers include other headers to improve compile times. Unfortunate that we have to add this one.

nmoucht replied:
I'm not sure there's a great way to avoid this unfortunately.

darinadler commented on Source/WebCore/css/ComputedStyleExtractor.cpp:
```diff
@@ -3207,6 +3209,36 @@ static Ref<CSSValue> valueForTimelineScopeNames(const Vector<AtomString>& names)
     return CSSValueList::createCommaSeparated(WTFMove(list));
 }
 
+static Ref<CSSValue> valueForSingleAnimationRange(const RenderStyle& style, const SingleTimelineRange& range, TimelineRangeStartOrEnd startOrEnd)
```
It's unfortunate to write new computed style extractor functions for new shorthands. We should be able to rely on the shorthand serializer instead at some point. Shorthands themselves don't really need computed values; the tests are checking the serialization, not the values. But maybe we aren't ready for that yet? Or maybe this isn't a shorthand?

nmoucht replied:
Yeah this is for the individual timeline ranges.

darinadler commented on Source/WebCore/css/ComputedStyleExtractor.cpp:
```diff
@@ -3207,6 +3209,36 @@ static Ref<CSSValue> valueForTimelineScopeNames(const Vector<AtomString>& names)
+    return CSSValueList::createSpaceSeparated(WTFMove(list));
+}
+
+static Ref<CSSValue> valueForAnimationRange(const RenderStyle& style)
```
t's unfortunate to write new computed style extractor functions for new shorthands. We should be able to rely on the shorthand serializer instead at some point. Shorthands themselves don't really need computed values; the tests are checking the serialization, not the values. But maybe we aren't ready for that yet? Or maybe this isn't a shorthand?

nmoucht replied:
@nt1m Any opinions here? Not sure why we haven't unified these two areas yet.

nt1m replied:
I think you can mark the property as usable with ShorthandSerializer here: https://searchfox.org/wubkat/rev/cde6cf2e21001cb4055a5f891ce3d7abe5eee3e9/Source/WebCore/css/CSSComputedStyleDeclaration.cpp#126-150

There's unfortunately a bunch of call sites that still require some code in ComputedStyleExtractor though, so I would just keep it as is.

darinadler commented on Source/WebCore/css/ShorthandSerializer.cpp:
```diff
@@ -1315,6 +1318,52 @@ String ShorthandSerializer::serializeTextWrap() const
+    if (auto* pair = dynamicDowncast<CSSValuePair>(rangeStart)) {
+        rangeStartID = pair->first().valueID();
+
+        if (SingleTimelineRange::isTimelineRangeDefaultValue(*dynamicDowncast<CSSPrimitiveValue>(pair->second()), TimelineRangeStartOrEnd::Start))
```
Using `*dynamicDowncast` is never correct. If we know the type we use `downcast`. If we don't know the type we need a null check after calling `dynamicDowncast`.

nmoucht replied:
Fixed.

darinadler commented on Source/WebCore/css/ShorthandSerializer.cpp:
```diff
@@ -1315,6 +1318,52 @@ String ShorthandSerializer::serializeTextWrap() const
+    }
+
+    if (auto* pair = dynamicDowncast<CSSValuePair>(rangeEnd)) {
+        if (SingleTimelineRange::isTimelineRangeDefaultValue(*dynamicDowncast<CSSPrimitiveValue>(pair->second()), TimelineRangeStartOrEnd::End) && pair->first().valueID() != rangeStartID)
```
Using `*dynamicDowncast` is never correct. If we know the type we use `downcast`. If we don't know the type we need a null check after calling `dynamicDowncast`.

nmoucht replied:
Fixed.

darinadler commented on Source/WebCore/css/ShorthandSerializer.cpp:
```diff
@@ -1315,6 +1318,52 @@ String ShorthandSerializer::serializeTextWrap() const
     return makeString(nameLiteral(mode), ' ', nameLiteral(style));
 }
 
+String ShorthandSerializer::serializeAnimationRange() const
```
The code in this function seems repetitive. We should look for helpers so we can refactor it to make it easier to read.

nmoucht replied:
Due to the special rules about omitting in certain cases, I'm not sure there's a good way to use existing serialization code here. I was able to refactor this to reduce duplicated code. Not sure if this solution is more readable though.

darinadler commented on Source/WebCore/css/ShorthandSerializer.cpp:
```diff
@@ -1315,6 +1318,52 @@ String ShorthandSerializer::serializeTextWrap() const
 
+String ShorthandSerializer::serializeAnimationRange() const
+{
+    auto& rangeStart = longhandValue(0);
```
This whole function is about a range so we can omit the word `range` from the names of all these local variables.

nmoucht replied:
Fixed.

darinadler commented on Source/WebCore/css/ShorthandSerializer.cpp:
```diff
@@ -1315,6 +1318,52 @@ String ShorthandSerializer::serializeTextWrap() const
+
+    auto rangeStartID = rangeStart.valueID();
+
+    auto rangeStartValue = emptyString();
```
Not great to have a local variable named `rangeStart` containing the range start value and a second one named `rangeStartValue`, since the names make it unclear how they are related. I would use the names `start` and `startString` or `start` and `serializedStart`.

nmoucht replied:
Fixed.

darinadler commented on Source/WebCore/css/parser/CSSPropertyParser.cpp:
```diff
@@ -2771,6 +2771,45 @@ bool CSSPropertyParser::consumeWhiteSpaceShorthand(bool important)
+
+bool CSSPropertyParser::consumeAnimationRangeShorthand(bool important)
+{
+    // TODO: handle animation range sequences
```
Our project uses `FIXME` not `TODO`. And we write comments sentence style with a period and a capital letter.

nmoucht replied:
Fixed.

darinadler commented on Source/WebCore/css/parser/CSSPropertyParser.cpp:
```diff
@@ -2771,6 +2771,45 @@ bool CSSPropertyParser::consumeWhiteSpaceShorthand(bool important)
+        // animation-range-end is set to that same and 100%. Otherwise, any omitted longhand is set to its initial value.
+
+        auto rangeEndValueForStartValue = [](const CSSValue& value) {
+            if (dynamicDowncast<CSSPrimitiveValue>(value) && SingleTimelineRange::isOffsetValue(*dynamicDowncast<CSSPrimitiveValue>(value)))
```
The point of `dynamicDowncast<>` is to use the return value, not to call it twice, so please put it in a local variable. If you aren't using the return value you use `is<>` instead. And if you aren't checking the nullity you use `downcast<>` instead.

nmoucht replied:
Fixed.

darinadler commented on Source/WebCore/css/parser/CSSPropertyParser.cpp:
```diff
@@ -2771,6 +2771,45 @@ bool CSSPropertyParser::consumeWhiteSpaceShorthand(bool important)
+bool CSSPropertyParser::consumeAnimationRangeShorthand(bool important)
+{
+    // TODO: handle animation range sequences
+    RefPtr animationRangeStart = consumeAnimationRangeStart(m_range, m_context);
```
This function is about the animation range so you don't need those words in the local variable names. This should just be named `start`.

nmoucht replied:
Fixed.

darinadler commented on Source/WebCore/css/parser/CSSPropertyParser.cpp:
```diff
@@ -2771,6 +2771,45 @@ bool CSSPropertyParser::consumeWhiteSpaceShorthand(bool important)
+    if (!animationRangeStart)
+        return false;
+
+    RefPtr<CSSValue> animationRangeEnd;
```
This function is about the animation range so you don't need those words in the local variable names. This should just be named `end`.

nmoucht replied:
Fixed.

darinadler commented on Source/WebCore/css/parser/CSSPropertyParser.cpp:
```diff
@@ -2771,6 +2771,45 @@ bool CSSPropertyParser::consumeWhiteSpaceShorthand(bool important)
+        if (RefPtr primitiveStart = dynamicDowncast<CSSPrimitiveValue>(animationRangeStart))
+            animationRangeEnd = rangeEndValueForStartValue(*primitiveStart);
+        else {
+            RefPtr pairStart = dynamicDowncast<CSSValuePair>(animationRangeStart);
```
If we know the type of something then we use `downcast<>`, not `dynamicDowncast<>` plus a handwritten assert.

nmoucht replied:
Fixed.

darinadler commented on Source/WebCore/css/parser/CSSPropertyParser.h:
```diff
@@ -156,6 +156,8 @@ class CSSPropertyParser {
     bool consumeTextWrapShorthand(bool important);
     bool consumeWhiteSpaceShorthand(bool important);
 
+    bool consumeAnimationRangeShorthand(bool important);
```
This class seems disorganized, with no obvious logic about what order and grouping there is for the function declarations. Not a new issue though. I would probably sort them alphabetically.

nmoucht replied:
Moved this to be with the other scroll driven animation consumers.

darinadler commented on Source/WebCore/css/parser/CSSPropertyParserConsumer+Timeline.cpp:
```diff
@@ -150,5 +152,43 @@ RefPtr<CSSValue> consumeViewTimelineInset(CSSParserTokenRange& range, const CSSP
+
+    auto id = range.peek().id();
+
+    if (id == CSSValueNormal)
```
This is unnecessarily inconsistent. We already consume an identifier after this. Why not check for `normal` there instead of peeking first?

nmoucht replied:
Fixed.

darinadler commented on Source/WebCore/css/parser/CSSPropertyParserConsumer+Timeline.h:
```diff
@@ -31,12 +31,15 @@ namespace WebCore {
 namespace CSSPropertyParserHelpers {
 
 // MARK: - Consumer functions
-
```
Why remove this blank line?

nmoucht replied:
Fixed.

darinadler commented on Source/WebCore/style/StyleBuilderConverter.h:
```diff
@@ -2187,6 +2191,36 @@ inline TimelineScope BuilderConverter::convertTimelineScope(BuilderState&, const
+        return { };
+
+    // <timeline-range-name> <length-percentage>
+    auto* primitiveValue = dynamicDowncast<CSSPrimitiveValue>(pair->second());
```
Instead of asserting use `downcast`.

nmoucht replied:
Fixed.

darinadler commented on Source/WebCore/animation/TimelineRange.h:
```diff
@@ -29,16 +29,30 @@
+
+    static bool isOffsetValue(const CSSPrimitiveValue&);
+
+    static Name timelineName(CSSValueID);
```
I know you made the function name shorter but just for the record, I still prefer the single word `name`.

nmoucht replied:
This was to avoid clashing with the member `name`.

darinadler commented on Source/WebCore/css/ComputedStyleExtractor.cpp:
```diff
@@ -86,11 +86,13 @@
 #include "ViewTimeline.h"
 #include "WebAnimationUtilities.h"
 
+
```
No reason to add this blank line.

nmoucht replied:
Fixed.

nt1m commented on Source/WebCore/style/StyleBuilderConverter.h:
```diff
@@ -2190,6 +2194,36 @@ inline TimelineScope BuilderConverter::convertTimelineScope(BuilderState&, const
+            return { SingleTimelineRange::Name::Omitted, convertLength(state, *primitiveValue) };
+        }
+        // <timeline-range-name> or Normal
+        return { SingleTimelineRange::timelineName(primitiveValue->valueID()), (startOrEnd == TimelineRangeStartOrEnd::Start ? WebCore::Length(0, LengthType::Percent) : WebCore::Length(100, LengthType::Percent)) };
```
You could probably drop `WebCore::` from Length?

nmoucht replied:
The compiler gets confused with WebCore::Style::Length here if I don't include it

nt1m commented on Source/WebCore/css/parser/CSSPropertyParser.cpp:
```diff
@@ -2772,6 +2772,44 @@ bool CSSPropertyParser::consumeWhiteSpaceShorthand(bool important)
+        if (RefPtr primitiveStart = dynamicDowncast<CSSPrimitiveValue>(start))
+            end = rangeEndValueForStartValue(*primitiveStart);
+        else {
+            RefPtr pairStart = downcast<CSSValuePair>(start);
```
Can we name these startPrimitiveValue / startPair? 

Putting start at the end makes it seems we're getting the start of the primitive value, rather than the primitive value of the start.

nmoucht replied:
Fixed.

nt1m commented on Source/WebCore/css/ShorthandSerializer.cpp:
```diff
@@ -1315,6 +1319,35 @@ String ShorthandSerializer::serializeTextWrap() const
 
+String ShorthandSerializer::serializeSingleAnimationRange(CSSValue& value, TimelineRangeStartOrEnd startOrEnd, CSSValueID startValueID) const
+{
+    if (auto* pair = dynamicDowncast<CSSValuePair>(value)) {
```
```suggestion
    if (RefPtr pair = dynamicDowncast<CSSValuePair>(value)) {
```

nmoucht replied:
Fixed.

nt1m commented on Source/WebCore/css/ShorthandSerializer.cpp:
```diff
@@ -1315,6 +1319,35 @@ String ShorthandSerializer::serializeTextWrap() const
+            return nameLiteral(pair->first().valueID());
+        return serializeLonghandValue(startOrEnd == TimelineRangeStartOrEnd::Start ? 0 : 1);
+    }
+    if (auto* primitiveValue = dynamicDowncast<CSSPrimitiveValue>(value)) {
```
```suggestion
    if (RefPtr primitiveValue = dynamicDowncast<CSSPrimitiveValue>(value)) {
```

nmoucht replied:
Fixed.

nt1m commented on Source/WebCore/css/ShorthandSerializer.cpp:
```diff
@@ -1315,6 +1319,35 @@ String ShorthandSerializer::serializeTextWrap() const
+String ShorthandSerializer::serializeSingleAnimationRange(CSSValue& value, TimelineRangeStartOrEnd startOrEnd, CSSValueID startValueID) const
+{
+    if (auto* pair = dynamicDowncast<CSSValuePair>(value)) {
+        if (SingleTimelineRange::isDefault(downcast<CSSPrimitiveValue>(pair->second()), TimelineRangeStartOrEnd::Start) && (startOrEnd == TimelineRangeStartOrEnd::Start || pair->first().valueID() != startValueID))
```
I think it would be worth splitting this into variables like for ComputedStyleExtractor, or at least adding some comments describing the different cases.

nmoucht replied:
Fixed.

nt1m commented on Source/WebCore/css/ShorthandSerializer.cpp:
```diff
@@ -1315,6 +1319,35 @@ String ShorthandSerializer::serializeTextWrap() const
+    if (auto* primitiveValue = dynamicDowncast<CSSPrimitiveValue>(value)) {
+        if (SingleTimelineRange::isOffsetValue(*primitiveValue))
+            return serializeLonghandValue(startOrEnd == TimelineRangeStartOrEnd::Start ? 0 : 1);
+        if (startOrEnd == TimelineRangeStartOrEnd::Start || (primitiveValue->valueID() != CSSValueNormal && primitiveValue->valueID() != startValueID))
```
ditto here.

nmoucht replied:
Fixed.

nt1m commented on Source/WebCore/css/parser/CSSPropertyParser.cpp:
```diff
@@ -2772,6 +2772,44 @@ bool CSSPropertyParser::consumeWhiteSpaceShorthand(bool important)
+        // animation-range-end is set to that same and 100%. Otherwise, any omitted longhand is set to its initial value.
+
+        auto rangeEndValueForStartValue = [](const CSSValue& value) {
+            if (is<CSSPrimitiveValue>(value) && SingleTimelineRange::isOffsetValue(downcast<CSSPrimitiveValue>(value)))
```
The common pattern is this:

```suggestion
            RefPtr primitiveValue = dynamicDowncast<CSSPrimitiveValue>(value);
            if (primitiveValue && SingleTimelineRange::isOffsetValue(primitiveValue))
```

nmoucht replied:
Fixed.

nt1m commented on Source/WebCore/css/ShorthandSerializer.cpp:
```diff
@@ -1315,6 +1319,35 @@ String ShorthandSerializer::serializeTextWrap() const
+String ShorthandSerializer::serializeAnimationRange() const
+{
+    auto& start = longhandValue(0);
+    auto startID = is<CSSValuePair>(start) ? downcast<CSSValuePair>(start).first().valueID() : start.valueID();
```
Similarly here, the common pattern is:
```suggestion
    RefPtr startPair = dynamicDowncast<CSSValuePair>(start);
    auto startID = startPair ? startPair.first().valueID() : start.valueID();
```

nmoucht replied:
Fixed.

nt1m commented on Source/WebCore/style/StyleBuilderConverter.h:
```diff
@@ -2190,6 +2194,36 @@ inline TimelineScope BuilderConverter::convertTimelineScope(BuilderState&, const
 
+inline SingleTimelineRange BuilderConverter::convertAnimationRange(BuilderState& state, const CSSValue& value, TimelineRangeStartOrEnd startOrEnd)
+{
+    if (auto* primitiveValue = dynamicDowncast<CSSPrimitiveValue>(value)) {
```
```suggestion
    if (RefPtr primitiveValue = dynamicDowncast<CSSPrimitiveValue>(value)) {
```

nmoucht replied:
Fixed.

nt1m commented on Source/WebCore/style/StyleBuilderConverter.h:
```diff
@@ -2190,6 +2194,36 @@ inline TimelineScope BuilderConverter::convertTimelineScope(BuilderState&, const
+        // <timeline-range-name> or Normal
+        return { SingleTimelineRange::timelineName(primitiveValue->valueID()), (startOrEnd == TimelineRangeStartOrEnd::Start ? WebCore::Length(0, LengthType::Percent) : WebCore::Length(100, LengthType::Percent)) };
+    }
+    auto* pair = dynamicDowncast<CSSValuePair>(value);
```
```suggestion
    RefPtr pair = dynamicDowncast<CSSValuePair>(value);
```

nmoucht replied:
Fixed.

nt1m reviewed (commented):
Looks generally good to go!

nt1m commented on Source/WebCore/animation/TimelineRange.h:
```diff
@@ -29,16 +29,30 @@
 
 namespace WebCore {
 
+enum class TimelineRangeStartOrEnd : bool { Start, End };
```
Very weak opinion, but should this be named TimelineRangeType / SingleTimelineRangeType / SingleTimelineRange::Type ? Wdyt ?

darinadler replied:
I don’t think "start" and "end" are types of ranges, so this shouldn't be named "range type".

Elsewhere I use words like "boundary", "edge", or "endpoint" for this concept. I could imagine naming the enumeration `TimelineRangeBoundary`. I find "x or y" to be awkward in type names, although it does happen.

nmoucht replied:
I'm not sure either are quite right. This is just to represent if the current timeline is for the starting range or ending range (to determine what the default offset should be). I do think it makes sense to net this in SingleTimelineRange, so made this SingleTimelineRange::StartOrEnd.

nt1m commented on Source/WebCore/css/parser/CSSPropertyParserConsumer+Timeline.cpp:
```diff
@@ -150,5 +152,41 @@ RefPtr<CSSValue> consumeViewTimelineInset(CSSParserTokenRange& range, const CSSP
+    return identMatches<CSSValueNormal, CSSValueCover, CSSValueContain, CSSValueEntry, CSSValueExit, CSSValueEntryCrossing, CSSValueExitCrossing>(id);
+}
+
+RefPtr<CSSValue> consumeAnimationRange(CSSParserTokenRange& range, const CSSParserContext& context, TimelineRangeStartOrEnd isStart)
```
```suggestion
RefPtr<CSSValue> consumeAnimationRange(CSSParserTokenRange& range, const CSSParserContext& context, TimelineRangeStartOrEnd startOrEnd)
```

nit: please be consistent with enum name, whatever it ends being

nmoucht replied:
Fixed.

nt1m reviewed (approved):
Seems fine, just wondering about the naming of `TimelineRangeStartOrEnd`

darinadler commented on Source/WebCore/animation/TimelineRange.h:
```diff
@@ -34,11 +34,24 @@ struct SingleTimelineRange {
+
+    bool operator==(const SingleTimelineRange& other) const = default;
+
+    enum class StartOrEnd : bool { Start, End };
```
I know we debated this already, but even after your explanation I probably would have named this `Type` instead of `StartOrEnd`.

nmoucht replied:
Fixed.

darinadler commented on Source/WebCore/animation/TimelineRange.h:
```diff
@@ -34,11 +34,24 @@ struct SingleTimelineRange {
     Name name { Name::Normal };
     Length offset;
+
+    bool operator==(const SingleTimelineRange& other) const = default;
```
Don’t need the word "other" here, by the way. I wouldn’t hold up the pull request just for this, though.

nmoucht committed:
[scroll-animations] Implement parsing support for animation-range and its longhand properties
https://bugs.webkit.org/show_bug.cgi?id=280862
rdar://137717240

Reviewed by Tim Nguyen and Darin Adler.

Implement parsing support for animation-range and its longhand properties. This needs a follow up
for sequences of animation ranges. There is also a failing subtest about determining if a calc
expression is a default offset value that needs some thought.

* Source/WebCore/Headers.cmake:
* Source/WebCore/animation/AnimationTimeline.h:
(WebCore::SingleTimelineRange::operator== const):
* Source/WebCore/animation/TimelineRangeOffset.h:
(WebCore::TimelineRangeOffset::operator== const):
* Source/WebCore/animation/WebAnimationTypes.h:
* Source/WebCore/css/CSSProperties.json:
* Source/WebCore/css/CSSValueKeywords.in:
* Source/WebCore/css/ComputedStyleExtractor.cpp:
(WebCore::valueForAnimationRange):
(WebCore::ComputedStyleExtractor::valueForPropertyInStyle const):
* Source/WebCore/css/parser/CSSPropertyParserConsumer+Timeline.cpp:
(WebCore::CSSPropertyParserHelpers::consumeAnimationRange):
* Source/WebCore/css/parser/CSSPropertyParserConsumer+Timeline.h:
* Source/WebCore/rendering/style/RenderStyle.h:
* Source/WebCore/rendering/style/RenderStyleInlines.h:
(WebCore::RenderStyle::initialAnimationRangeStart):
(WebCore::RenderStyle::initialAnimationRangeEnd):
(WebCore::RenderStyle::animationRangeStart const):
(WebCore::RenderStyle::animationRangeEnd const):
* Source/WebCore/rendering/style/RenderStyleSetters.h:
(WebCore::RenderStyle::setAnimationRangeStart):
(WebCore::RenderStyle::setAnimationRangeEnd):
* Source/WebCore/rendering/style/StyleRareNonInheritedData.cpp:
(WebCore::StyleRareNonInheritedData::StyleRareNonInheritedData):
(WebCore::StyleRareNonInheritedData::operator== const):
* Source/WebCore/rendering/style/StyleRareNonInheritedData.h:
* Source/WebCore/style/StyleBuilderConverter.h:
(WebCore::Style::BuilderConverter::convertAnimationRangeStart):
(WebCore::Style::BuilderConverter::convertAnimationRangeEnd):
(WebCore::Style::BuilderConverter::convertAnimationRange):

Canonical link: https://commits.webkit.org/285135@main
